### ISSUE TITLE
Compile errors using @bitCast u8 slice to f32

### ISSUE THREAD

**Issue by jeffkdev** [ziglang/zig#4254]
Shared main function for the below tests:
```zig
pub fn main() void {
    const float_array = [1]f32{1234.567};
    const float_as_bytes = @sliceToBytes(float_array[0..]);
    const result = readF32(float_as_bytes);
    @import("std").debug.warn("result = {}\n", .{result});
}
```
Function 1:
running "zig build-exe scratch.zig" on Windows command line ouputs no errors and no executable. If I run it in my dev environment in VScode I get the error: "The following command exited with error code 5:" [full command input]
```zig
pub fn readF32(bytes: []const u8) f32 {
    const array=[4]u8 {
        bytes[0],
        bytes[1],
        bytes[2],
        bytes[3],
    };
    return @bitCast(f32, array[0..]);
}
```

Function 2: This gives a compile error:
.\scratch.zig:3:14: error: cannot store runtime value in compile time variable
        bytes[0],
             ^
```zig
pub fn readF32(bytes: []const u8) f32 {
    return @bitCast(f32, [4]u8 {
        bytes[0],
        bytes[1],
        bytes[2],
        bytes[3],
    });
}
```

Function 3: For reference, this compiles and runs successfully:
result = 1.23456701e+03
```zig
pub fn readF32(bytes: []const u8) f32 {
    const array=[4]u8 {
        bytes[0],
        bytes[1],
        bytes[2],
        bytes[3],
    };
    return @bitCast(f32, array);
}
```

The actual use case involves reading the bytes input array at different offsets, I simplified to the example functions above just to read one float from the buffer.

0.5.0 +a6f6d8d2f

**Comment by LemonBoy** [comment]
> Function 1:

Slices are nothing but small structures containing a pointer + a length, you're possibly casting one of the two here to a `f32` and that's not what you want.

> Function 2:

That looks like a bug

> Function 3:

Now you're casting the whole array (note that arrays != slices) as a f32. `bytesToSlice` is also useful here if you want something easier to read or have to convert more than a single `f32` at once.

**Comment by jeffkdev** [comment]
> Slices are nothing but small structures containing a pointer + a length, you're possibly casting one of the two here to a f32 and that's not what you want.

Yeah, I wasn't very clear what I meant on that one. I was trying to say Function 1 should give some sort of compiler error since it silently fails right now. Not that it should give the same result as Function 2 and 3. For example this gives an error:
```zig
const too_small : u1 = 1;
var value :u32 = @bitCast(u32, too_small);
```
```
error: destination type 'u32' has size 4 but source type 'u1' has size 1
    var value :u32 = @bitCast(u32, too_small);
```
In this case the slice shouldn't be able to fit in in a f32. Should I open a new issue for this one and modify the OP to just show Function 2?

**Comment by andrewrk** [comment]
Testing with latest master 

----

> Function 1:

```
./test2.zig:8:31: error: destination type 'f32' has size 4 but source type '[]const u8' has size 16
    return @bitCast(f32, array[0..]);
                              ^
```

:heavy_check_mark:  correct compile error

> Function 2:

```
result = 1.23456701e+03
```

:heavy_check_mark:  correctly runs and gives output

> Function 3

```
result = 1.23456701e+03
```

:heavy_check_mark:  correctly runs and gives output

----

I believe the bug was solved by #4152, which introduced several new `@bitCast` tests, so I'm sure this is covered.
